src/main/java/com/feedzai/commons/sql/abstraction/engine/AbstractTranslator.java

```diff
@@ -492,6 +493,54 @@ public String translate(final Values.Row row) {
         return row.isEnclosed() ? "(" + translation + ")" : translation;
     }
 
+    /**
+     * Transform values' rows into a union.
+     *
+     * @param rows the values' rows
+     * @return the resulting union
+     */
+    protected Union rowsToUnion(final List<Expression> rows) {
+        return toUnionTree(rows);
+    }
+
+    /**
+     * Create an union in form of a binary tree from a list of rows.
+     * The tree shape is to prevent stack overflow on some
+     * database engines when the list of rows is too big.
+     *
+     * @param rows the list of rows.
+     * @return an union in form of a binary tree.
+     */
+    private Union toUnionTree(final List<Expression> rows) {
+
+        List<Expression> rowsWithSelect = new ArrayList<>(rows);
+        
+        while (rowsWithSelect.size() > 2) {
+            final List<Expression> newRowsWithSelect = new ArrayList<>();
+
+            // Put the first and second rowsWithSelect together forming an UNION ALL.
+            // Then the third and forth, and so on.
+            for (int i = 1; i < rowsWithSelect.size(); i++) {
+                if ((i - 1) % 2 == 0) {
```

Comment:
Why do this instead of `i+=2` in the above line?  
Am I missing something?

Reply:
;) You're right